circuit_passes/src/bucket_interpreter/env/function_env.rs

```diff
@@ -0,0 +1,235 @@
+    //  check the base Env because that lookup would further increase the Rust stack height
+    //  while doing the check that is intended to prevent Rust stack exhaustion.
+    call_stack: CallStack,
+}
+
+impl Display for FunctionEnvData<'_> {
+    fn fmt(&self, f: &mut Formatter<'_>) -> std::fmt::Result {
+        if PRINT_ENV_SORTED {
+            write!(
+                f,
+                "FunctionEnv{{\n  ctx = {:?}\n  vars = {:?}\n}}",
+                self.caller_stack,
+                sort(&self.vars, std::convert::identity),
+            )
+        } else {
+            write!(
+                f,
+                "FunctionEnv{{\n  ctx = {:?}\n  vars = {:?}\n}}",
+                self.caller_stack, self.vars,
+            )
+        }
+    }
+}
+
+impl LibraryAccess for FunctionEnvData<'_> {
+    fn get_function(&self, name: &String) -> Ref<FunctionCode> {
+        self.libs.get_function(name)
+    }
+
+    fn get_template(&self, name: &String) -> Ref<TemplateCode> {
+        self.libs.get_template(name)
+    }
+}
+
+impl<'a> FunctionEnvData<'a> {
+    pub fn new(
+        base: Env<'a>,
+        caller: &BucketId,
+        call_stack: CallStack,
+        libs: &'a dyn LibraryAccess,
+    ) -> Self {
+        let mut caller_stack = base.get_caller_stack().to_vec();
+        caller_stack.push(*caller);
+        FunctionEnvData { caller_stack, call_stack, vars: Default::default(), libs }
+    }
+
+    // READ OPERATIONS
+    pub fn get_context_kind(&self) -> EnvContextKind {
+        EnvContextKind::SourceFunction
+    }
+
+    pub fn append_stack_if_safe_to_interpret(
+        &self,
+        new_frame: CallStackFrame,
+    ) -> Option<CallStack> {
+        if !self.call_stack.contains(&new_frame) && self.call_stack.depth() <= CALL_STACK_LIMIT {
+            Some(self.call_stack.clone().push(new_frame))
+        } else {
+            None
+        }
+    }
+
+    pub fn get_caller_stack(&self) -> &[BucketId] {
+        &self.caller_stack
+    }
+
+    pub fn get_var(&self, idx: usize) -> Value {
+        self.vars.get(&idx).unwrap_or_default().clone()
+    }
+
+    pub fn get_signal(&self, _idx: usize) -> Value {
+        // There are no signals in source functions.
+        unreachable!()
+    }
+
+    pub fn get_subcmp_signal(&self, _subcmp_idx: usize, _signal_idx: usize) -> Value {
+        // There are no components in source functions.
+        unreachable!()
+    }
+
+    pub fn get_subcmp_name(&self, _subcmp_idx: usize) -> &String {
+        // There are no components in source functions.
+        unreachable!()
+    }
+
+    pub fn get_subcmp_template_id(&self, _subcmp_idx: usize) -> usize {
+        // There are no components in source functions.
+        unreachable!()
+    }
+
+    pub fn get_subcmp_counter(&self, _subcmp_idx: usize) -> Value {
+        // There are no components in source functions.
+        unreachable!()
+    }
```

Comment:
Shouldn't we remove these functions as well?

Reply:
The way I simplified implementation of the top Env requires them.

Reply:
I didn't create a Trait for these specific types of Env but I used essentially that pattern and this function would be required by the trait.